### PR TITLE
[ADD] Bitwise operation to speed up

### PR DESCRIPTION
**Pull request by agavrel**
Negative number will gives its true value, positive number will be equal to 0. MUL always better than JMP.
Regards,

### CODE REVIEW

**Comment by cjihrig** [comment]
Unless this is fixing a bug, I'd rather stick to more readable code, and let the compiler handle optimizations.

**Comment by gireeshpunathil** [comment]
This scenario seem to be very unique in that a compiler is not going to replace the jmp, even with -O5 for the existing code, while the proposed change make it happen.

```c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {

  int x = atoi(argv[1]);
  if(x >= 0)
    printf("%d\n", 0);
  else
    printf("%d\n", x);
}
```
`
$ gcc -O5 -g foo.c
`

```javascript
(gdb) x/20i main+20
   0x400524 <main+20>:	test   eax,eax
   0x400526 <main+22>:	js     0x400540 <main+48>
   0x400528 <main+24>:	xor    esi,esi
   0x40052a <main+26>:	mov    edi,0x400658
   0x40052f <main+31>:	xor    eax,eax
   0x400531 <main+33>:	add    rsp,0x8
   0x400535 <main+37>:	jmp    0x4003f0 <printf@plt>
   0x40053a <main+42>:	nop    WORD PTR [rax+rax*1+0x0]
   0x400540 <main+48>:	mov    esi,eax
   0x400542 <main+50>:	mov    edi,0x400658
   0x400547 <main+55>:	xor    eax,eax
   0x400549 <main+57>:	add    rsp,0x8
   0x40054d <main+61>:	jmp    0x4003f0 <printf@plt>
   0x400552:	nop
   0x400553:	nop
```

with the proposal:

```c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
  int x = atoi(argv[1]);
  int y = ((x >> 31) * x) * -1;
  printf("%d\n", y);
}
```

```javascript
(gdb) x/20i main+20
   0x400524 <main+20>:	mov    esi,eax
   0x400526 <main+22>:	mov    edi,0x400638
   0x40052b <main+27>:	add    rsp,0x8
   0x40052f <main+31>:	shr    esi,0x1f
   0x400532 <main+34>:	imul   esi,eax
   0x400535 <main+37>:	xor    eax,eax
   0x400537 <main+39>:	jmp    0x4003f0 <printf@plt>
```

So clearly there is a benefit, the extent of which will be visible based on the intensity of usage in the app.

So @cjihrig - PTAL! Probably with sufficient comments in the surroundings to aid readability?

File: src/unix/udp.c [hunk 1]
```diff
@@ -115,10 +117,12 @@ static void uv__udp_run_completed(uv_udp_t* handle) {
     /* req->status >= 0 == bytes written
      * req->status <  0 == errno
      */
+    req->send_cb(req, (req->status >> (CHAR_BIT * sizeof(integer) - 1)) * req->status);
```
Comment (gireeshpunathil) [inline src/unix/udp.c]: what is CHAR_BIT? You meant 8 I guess.
what is integer? You meant `int` I guess.
Also, I guess you need to multiply the whole expression with -1, to bring back the sign.

**Comment by bnoordhuis** [comment]
It's an optimization that is unlikely to be noticeable in the grand scheme of things and it hurts legibility, never mind that it looks like it doesn't compile.  So, -1.

I'll close this out but thanks for the PR.

File: src/unix/udp.c (hunk 1 above)
Comment (agavrel) [reply src/unix/udp.c]: Thank you for the tests anyways and agree with all of you ;)

Comment (agavrel) [reply src/unix/udp.c]: FIXED;
req->send_cb(req, (req->status >> (CHAR_BIT * sizeof(int) - 1)) * req->status * -1);

Comment (agavrel) [reply src/unix/udp.c]: hello Gireesh, could you tell me how you do this:

$ gcc -O5 -g foo.c

(gdb) x/20i main+20
   0x400524 <main+20>:	test   eax,eax
   0x400526 <main+22>:	js     0x400540 <main+48>
   0x400528 <main+24>:	xor    esi,esi
   0x40052a <main+26>:	mov    edi,0x400658
   0x40052f <main+31>:	xor    eax,eax
   0x400531 <main+33>:	add    rsp,0x8
   0x400535 <main+37>:	jmp    0x4003f0 <printf@plt>
   0x40053a <main+42>:	nop    WORD PTR [rax+rax*1+0x0]
   0x400540 <main+48>:	mov    esi,eax
   0x400542 <main+50>:	mov    edi,0x400658
   0x400547 <main+55>:	xor    eax,eax
   0x400549 <main+57>:	add    rsp,0x8
   0x40054d <main+61>:	jmp    0x4003f0 <printf@plt>
   0x400552:	nop
   0x400553:	nop

to check the number of cycles ?

Comment (gireeshpunathil) [reply src/unix/udp.c]: sorry, I didn't follow you. Can you please clarify the question?

Comment (agavrel) [reply src/unix/udp.c]: np; how can you vizualize 'main+61' (61 being the cycle or latency I guess?) with gdb

Comment (gireeshpunathil) [reply src/unix/udp.c]: Still not clear on the question, but let me try:

```c
  if(x >= 0)
    printf("%d\n", 0);
  else
    printf("%d\n", x);
```

```asm
   0x400524 <main+20>:	test   eax,eax                    # x == 0 ?
   0x400526 <main+22>:	js     0x400540 <main+48>         # if -ve, go to 0x400540 (printf("%d\n", 0);)
   0x400528 <main+24>:	xor    esi,esi                    # clear esi
   0x40052a <main+26>:	mov    edi,0x400658               # load format specifier, "%d\n"
   0x40052f <main+31>:	xor    eax,eax                    # load param, 0
   0x400531 <main+33>:	add    rsp,0x8                    # push down stack ptr, to mimic a call
   0x400535 <main+37>:	jmp    0x4003f0 <printf@plt>      # invoke printf, and return directly to the caller.
   0x40053a <main+42>:	nop    WORD PTR [rax+rax*1+0x0]   # NOOP
   0x400540 <main+48>:	mov    esi,eax                    # x is -ve, load it into esi
   0x400542 <main+50>:	mov    edi,0x400658               # load format specifier, "%d\n"
   0x400547 <main+55>:	xor    eax,eax                    # clear eax
   0x400549 <main+57>:	add    rsp,0x8                    # push down stack ptr
   0x40054d <main+61>:	jmp    0x4003f0 <printf@plt>      # invoke printf
```

the only way to reach `<main+61>` is through `test` being true, and then jumping to `0x400540 <main+48>`
in other words, both the `if` and `else` block are reachable through the test conditional only.
You can step debug with gdb and see these values yourself.

Does it make things clear?

Comment (agavrel) [reply src/unix/udp.c]: really sorry but my question is only about the +31 +33 etc. just after main. how do you get these figures? (<main+37> etc)

Comment (gireeshpunathil) [reply src/unix/udp.c]: ah, ok - here is the steps. Detailed online documentation can be found [here](https://ftp.gnu.org/old-gnu/Manuals/gdb/html_node/gdb_toc.html)

```
#  gcc -O5 -g foo.c -m32
#  gdb a.out
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-75.el6)
Copyright (C) 2010 Free Software Foundation, Inc.
Reading symbols from /home/gireesh/a.out...done.
(gdb) break main
Breakpoint 1 at 0x8048449: file foo.c, line 7.
(gdb) run -3
Starting program: /home/gireesh/a.out -3

Breakpoint 1, main (argc=2, argv=0xffffd594) at foo.c:7
7	  int x = atoi(argv[1]);
(gdb) set disassembly-flavor intel 
(gdb) x/10i main
   0x8048440 <main>:	push   ebp
   0x8048441 <main+1>:	mov    ebp,esp
   0x8048443 <main+3>:	and    esp,0xfffffff0
   0x8048446 <main+6>:	sub    esp,0x10
=> 0x8048449 <main+9>:	mov    eax,DWORD PTR [ebp+0xc]
   0x804844c <main+12>:	mov    DWORD PTR [esp+0x8],0xa
   0x8048454 <main+20>:	mov    DWORD PTR [esp+0x4],0x0
   0x804845c <main+28>:	mov    eax,DWORD PTR [eax+0x4]
   0x804845f <main+31>:	mov    DWORD PTR [esp],eax
   0x8048462 <main+34>:	call   0x8048314 <strtol@plt>
(gdb) 
```

Comment (agavrel) [reply src/unix/udp.c]: thank you so much ;)

Comment (gireeshpunathil) [reply src/unix/udp.c]: those are neither cycles nor latencies - they are the integral offset of the code from the beginning of the subroutine.  From the above dump, the code starts at address `0x8048440` (main+0) and the first instruction is one byte long, so the next one starts at `0x8048441` which  is (main+1) and so on.

Comment (agavrel) [reply src/unix/udp.c]: ha wow !! Thank you for the clarification...  -so it means that my code decreased the size but nothing about cycle speed in your test ? also I could not compile with -O5, only O3 and Ofast but I will google it up. Thanks a lot !!

**Pull request closed without merging**